guides/authentication.md

```diff
@@ -0,0 +1,210 @@
+
+## Access tokens
+
+The solution to avoid using api keys is to use access tokens. This is short living tokens which grants the user access to CDF. The application get an access token by asking the user to login to the IdP-provier (Google, Active Directory etc) of a CDF project.
+
+## Authentication flow
+
+This describes the flow for the user using your application:
+1. The user is using your app on this url: `https://my-app.com/some/path?someQuery=someValue`
+2. The SDK redirect the user to the project's IdP provider
+3. After a successful log in the browser will redirect the user back to the url `https://my-app.com/some/path?someQuery=someValue&access_token=someToken&id_token=someToken`
+4. The app will initialize again and `createClientWithOAuth` will be called again
+5. The SDK now sees the query parameters `access_token` and `id_token` in the URL. It will parse them and remove them from the URL
+6. The SDK is now authenticated (using the access token from the URL)
+
+The access token will expire after some time and result in the user get `401` from CDF again. A new authentication process will be triggered by the SDK.
+
+## How to authenticate with the SDK?
+
+There are two ways to authenticate using the SDK:
+
+- [With redirect](#authentication-with-redirects)
+- [With popups](#authentication-with-popups)
+
+### Authentication with redirects
+
+When doing authentication with redirects the currect browser window will be redirected to the IdP-provider for the user to log in. After login the same browser window will be redirect back to your application.
+
+> You can find a running example application using redirects [here](../samples/react/authentication/src/App.js).
+
+#### Simple example
+
+```js
+import * as sdk from '@cognite/sdk';
+const client = sdk.createClientWithOAuth({
+  project: 'YOUR PROJECT NAME HERE',
+  // default is redirect
+  // but can be explicitly specified:
+  onAuthenticate: sdk.REDIRECT,
+});
+
+// then use the SDK:
+const assets = await client.assets.retrive({ id: 23232789217132 });
+```
+
+The first time this will run the user will get a `401`-response from CDF in the first call to `client.assets`. This will trigger the SDK to perform a redirect of the browser window to authenticate.
+
+The second time `client.assets` is called the SDK will be authenticated and the call will succeed.
+
+#### Customize redirect URL
+
+If you want a different redirect url back to your app after a successful / unsuccessful login you can implement this:
+
+```js
+import * as sdk from '@cognite/sdk';
+const client = sdk.createClientWithOAuth({
+  project: 'YOUR PROJECT NAME HERE',
+  onAuthenticate: login => {
+    login.redirect({
+      redirectUrl: 'https://my-app.com/successful-login',
+      errorRedirectUrl: 'https://my-app.com/unsuccessful-login', // this one is optional. Will default to redirectUrl
```

Comment:
PLEASE ENCOURAGE AN errorRedirectUrl. it is much harder to debug when things keep looping and nobody realizes they have the same destination (at least if the destination doesn't handle errors...)